consensus/istanbul/core/roundchange.go

```diff
@@ -376,29 +379,42 @@ func (rcs *roundChangeSet) String() string {
 		latestRoundForValStr = append(latestRoundForValStr, fmt.Sprintf("%v: %v", addr.String(), r))
 	}
 
-	return fmt.Sprintf("RCS len=%v mode_round=%v mode_round_len=%v unique_rounds=<%v> %v",
+	return fmt.Sprintf("RCS len=%v mode_round=%v mode_round_len=%v unique_rounds=%v %v",
 		len(rcs.latestRoundForVal),
 		modeRound,
 		modeRoundSize,
 		len(rcs.msgsForRound),
 		strings.Join(msgsForRoundStr, ", "))
 }
 
-// Gets a round change certificate for a specific round.
-func (rcs *roundChangeSet) getCertificate(r *big.Int, quorumSize int) (istanbul.RoundChangeCertificate, error) {
+// Gets a round change certificate for a specific round. Includes all messages of that round or later.
+// If the total is less than quorumSize, returns an empty cert and errFailedCreateRoundChangeCertificate.
+func (rcs *roundChangeSet) getCertificate(minRound *big.Int, quorumSize int) (istanbul.RoundChangeCertificate, error) {
 	rcs.mu.Lock()
 	defer rcs.mu.Unlock()
 
-	round := r.Uint64()
-	if rcs.msgsForRound[round] != nil && rcs.msgsForRound[round].Size() >= quorumSize {
-		messages := make([]istanbul.Message, rcs.msgsForRound[round].Size())
-		for i, message := range rcs.msgsForRound[round].Values() {
-			messages[i] = *message
+	// Sort rounds descending
+	var sortedRounds []uint64
+	for r := range rcs.msgsForRound {
+		sortedRounds = append(sortedRounds, r)
+	}
+	sort.Slice(sortedRounds, func(i, j int) bool { return sortedRounds[i] > sortedRounds[j] })
+
+	var messages []istanbul.Message
+	for _, r := range sortedRounds {
+		if r < minRound.Uint64() {
+			break
 		}
+		for _, message := range rcs.msgsForRound[r].Values() {
+			messages = append(messages, *message)
```

Comment:
Should we break here if `len(messages) == quorumSize`?

Reply:
Yes, I believe this is valid. Changed so that we never include more than `quorum` messages.